common/src/main/java/com/microsoft/identity/common/internal/authorities/AuthorityDeserializer.java

```diff
@@ -48,9 +51,14 @@ public Authority deserialize(JsonElement json, Type typeOfT, JsonDeserialization
                             TAG + methodName,
                             "Type: AAD"
                     );
-                    AzureActiveDirectoryAuthority aadAuthority = context.deserialize(authorityObject, AzureActiveDirectoryAuthority.class);
-                    if (aadAuthority != null && aadAuthority.mAudience != null) {
-                        aadAuthority.mAudience.setCloudUrl(aadAuthority.mAuthorityUrl);
+                    final AzureActiveDirectoryAuthority aadAuthority = context.deserialize(authorityObject, AzureActiveDirectoryAuthority.class);
+                    if (aadAuthority != null && aadAuthority.mAudience != null && aadAuthority.mAuthorityUrl != null) {
+                        final Uri uri = Uri.parse(aadAuthority.mAuthorityUrl);
+                        aadAuthority.mAudience.setCloudUrl(uri.getScheme() + "://" + uri.getHost());
+                        // Try to extract tenant ID from the authority link.
+                        if (!TextUtils.isEmpty(uri.getLastPathSegment())) {
+                            aadAuthority.mAudience.setTenantId(uri.getLastPathSegment());
+                        }
```

Comment:
Looks like we are just setting the tenant id on the audience -- whatever audience was created; which may not be correct in the first place i.e. not having any audience in the config defaults to `AllAccounts`. I think we need to make sure the the audience on the Authority is of the correct type.

Reply:
In this case, audience won't be provided. 

i.e. 
```
"authorities" : [
    {
      "type": "AAD",
      "authority_url": "https://login.microsoftonline.com/rapong.onmicrosoft.com"
    }
  ]
```

We wouldn't be able to identify the audience type in the first place.

Reply:
(both tenant name and tenant id work interchangeably)

Reply:
We should be able to by splitting the URL in parts and checking the last segment and create the following mapping:

- /common - AllAccounts
- /organizations - AnyOrganizationalAccount
- /consumers - AnyPersonalAccount
- /tenantId (or tenantName) - AccountsInOneOrganization

and then we can initialize the authority object with the correct audience

Reply:
ah ok. makes sense. will do.